### PR TITLE
Implement BSD Service

### PR DESCRIPTION
**Pull request by AcK77**
- Implementation of bsd:s & bsd:u.
- Adding an EndianSwap class.

### CODE REVIEW

File: Ryujinx.Core/OsHle/Services/Bsd/ServiceBsd.cs [hunk 1]
```diff
@@ -52,45 +77,356 @@ public long Initialize(ServiceCtx Context)
...
+            return 0;
+        }
+
+        //(u32 socket, u32 flags, buffer<i8, 0x21, 0>) -> (i32 ret, u32 bsd_errno)
+        public long Send(ServiceCtx Context)
+        {
+            int SocketId = Context.RequestData.ReadInt32();
+            int SocketFlags = Context.RequestData.ReadInt32();
+
+            byte[] SendedBuffer = AMemoryHelper.ReadBytes(Context.Memory, Context.Request.SendBuff[0].Position, (int)Context.Request.SendBuff[0].Size);
```
Comment (gdkchan) [inline Ryujinx.Core/OsHle/Services/Bsd/ServiceBsd.cs]: SentBuffer, not SendedBuffer

File: Ryujinx.Core/OsHle/Services/Bsd/ServiceBsd.cs [hunk 2]
```diff
@@ -52,45 +77,356 @@ public long Initialize(ServiceCtx Context)
...
+                }
+
+                Context.ResponseData.Write(Sockets.Count - 1);
+                Context.ResponseData.Write(0);
+                Context.ResponseData.Write(8); //Size of MS.
+            }
+            else
+            {
+                Context.ResponseData.Write(-1);
+                Context.ResponseData.Write(60); //ETIMEDOUT
```
Comment (gdkchan) [inline Ryujinx.Core/OsHle/Services/Bsd/ServiceBsd.cs]: Can you add a enum with those error codes? This eliminates the need for comments too.

File: Ryujinx.Core/OsHle/Services/Bsd/ServiceBsd.cs [hunk 3]
```diff
@@ -52,45 +77,356 @@ public long Initialize(ServiceCtx Context)
...
+                    Writer.Write(byte.Parse(IpAdress[1]));
+                    Writer.Write(byte.Parse(IpAdress[2]));
+                    Writer.Write(byte.Parse(IpAdress[3]));
+
+                    AMemoryHelper.WriteBytes(Context.Memory, AddrBufferPtr, MS.ToArray());
+                }
+
+                Context.ResponseData.Write(Sockets.Count - 1);
+                Context.ResponseData.Write(0);
+                Context.ResponseData.Write(8); //Size of MS.
```
Comment (gdkchan) [inline Ryujinx.Core/OsHle/Services/Bsd/ServiceBsd.cs]: Move this inside the using and use `MS.Length` instead?

File: Ryujinx.Core/OsHle/Services/Bsd/ServiceBsd.cs [hunk 4]
```diff
@@ -52,45 +77,356 @@ public long Initialize(ServiceCtx Context)
...
+
+        //(u32 socket) -> (i32 ret, u32 bsd_errno)
+        public long Close(ServiceCtx Context)
+        {
+            int SocketId = Context.RequestData.ReadInt32();
+
+            try
+            {
+                Sockets[SocketId].Handle.Close();
+                Sockets[SocketId] = null; //Don't Remove() because other Sockets will change of Id.
```
Comment (gdkchan) [inline Ryujinx.Core/OsHle/Services/Bsd/ServiceBsd.cs]: Comment is confusing IMO.

File: Ryujinx.Core/OsHle/Services/Bsd/ServiceBsd.cs [hunk 5]
```diff
@@ -52,45 +77,356 @@ public long Initialize(ServiceCtx Context)
...
+            int PollCount = Context.RequestData.ReadInt32();
+            int TimeOut = Context.RequestData.ReadInt32();
+
+            //https://github.com/torvalds/linux/blob/master/include/uapi/asm-generic/poll.h
+            //https://msdn.microsoft.com/fr-fr/library/system.net.sockets.socket.poll(v=vs.110).aspx
+            //https://github.com/switchbrew/libnx/blob/e0457c4534b3c37426d83e1a620f82cb28c3b528/nx/source/services/bsd.c#L343
+            //https://github.com/TuxSH/ftpd/blob/switch_pr/source/ftp.c#L1634
+            //https://linux.die.net/man/2/poll
+
+            byte[] SendedBuffer = AMemoryHelper.ReadBytes(Context.Memory, Context.Request.SendBuff[0].Position, (int)Context.Request.SendBuff[0].Size);
```
Comment (gdkchan) [inline Ryujinx.Core/OsHle/Services/Bsd/ServiceBsd.cs]: SentBuffer

File: Ryujinx.Core/OsHle/Services/Bsd/ServiceBsd.cs [hunk 6]
```diff
@@ -52,45 +77,356 @@ public long Initialize(ServiceCtx Context)
...
+            int SocketId = BitConverter.ToInt32(SendedBuffer, 0);
```
Comment (gdkchan) [inline Ryujinx.Core/OsHle/Services/Bsd/ServiceBsd.cs]: Please don't use BitConverter.ToInt* methods, they are machine endian dependent.

File: Ryujinx.Core/OsHle/Services/Bsd/ServiceBsd.cs [hunk 7]
```diff
@@ -1,41 +1,66 @@
...
+using System;
 using System.Collections.Generic;
+using System.IO;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading.Tasks;
 
 namespace Ryujinx.Core.OsHle.IpcServices.Bsd
 {
+    class SocketBSD
```
Comment (gdkchan) [inline Ryujinx.Core/OsHle/Services/Bsd/ServiceBsd.cs]: SocketBSD -> SocketBsd to be consistent.

File: Ryujinx.Core/OsHle/Services/Bsd/ServiceBsd.cs [hunk 8]
```diff
@@ -52,45 +77,356 @@ public long Initialize(ServiceCtx Context)
...
+
+            return 0;
+        }
+
+        //(u32 socket, buffer<sockaddr, 0x21, 0>) -> (i32 ret, u32 bsd_errno)
+        public long Bind(ServiceCtx Context)
+        {
+            int SocketId = Context.RequestData.ReadInt32();
+
+            byte[] AddressBuffer = AMemoryHelper.ReadBytes(Context.Memory, Context.Request.SendBuff[0].Position, (int)Context.Request.SendBuff[0].Size);
```
Comment (gdkchan) [inline Ryujinx.Core/OsHle/Services/Bsd/ServiceBsd.cs]: Can you break that into multiple lines, this line is too long.

**Comment by gdkchan** [comment]
lgtm, thanks!

File: Ryujinx.Core/OsHle/Services/Bsd/ServiceBsd.cs (hunk 8 above)
Comment (marcussacana) [reply Ryujinx.Core/OsHle/Services/Bsd/ServiceBsd.cs]: at line 321 too

**Pull request merged**